index.js

```diff
@@ -1,135 +1,286 @@
+ * @param {string}        s2                              Second comparison string
+ * @param {Object}        options                         Options hash
+ * @param {bool}          options.calculateEditDistance   If true, return the editDistance of transformed strings with the isEquivalent boolean
+ * @param {bool}          options.giveReasons             When true, include the reason(s) why the strings aren't equivalent.
+ *
+ * @return {Object} Returns an object with the following top-level
+ *                  properties:
+ *                  - isEquivalent
+ *                  - editDistance (optional)
+ *                  - reasons[] (optional)
+ */
+
+Equivalency.prototype.equivalent = function(s1, s2, options = null) {
+  // Ensure identity is the final and only the final rlue.
+  if (
+    this._ruleList.length === 0 ||
+    this._ruleList[this._ruleList.length - 1].rule !== identityRule
+  )
+    this._ruleList.push({ rule: identityRule, matters: true });
+
+  const [finalMap, ruleFns] = Equivalency._collapseRules(this._ruleList);
+
+  const { isEquivalent, s1prime, s2prime } = Equivalency._compareWithRules(
+    s1,
+    s2,
+    finalMap,
+    ruleFns
+  );
+  let results = { isEquivalent: isEquivalent };
 
-    return results;
+  if (options && options.calculateEditDistance) {
+    const editDistance = dl(s1prime, s2prime);
+    results.editDistance = editDistance.steps;
   }
 
-  rules() {
-    return this._rules;
+  if (options && options.giveReasons) {
+    if (isEquivalent) {
+      results.reasons = [];
+    } else {
+      const reasons = [];
+      const indices = [];
+
+      // First, make all the matters rules doesntMatter. If it still fails,
+      // then push identity and we're done.
+      const indexesOfRulesThatMatter = this._ruleList
+        .map((r, idx) => idx)
+        .filter(idx => this._ruleList[idx].matters);
+
+      const allDoesntMatterRules = this._ruleList.map((rule, idx) => {
+        if (
+          indexesOfRulesThatMatter.indexOf(idx) > -1 &&
+          rule.rule !== identityRule
+        ) {
+          rule.matters = false;
+        }
+        return rule;
+      });
+      const [finalMap, ruleFns] = Equivalency._collapseRules(
+        allDoesntMatterRules
+      );
+      const { isEquivalent } = Equivalency._compareWithRules(
+        s1,
+        s2,
+        finalMap,
+        ruleFns
+      );
+      if (!isEquivalent) {
+        reasons.push(identityRule);
+      }
+
+      // restore
+      this._ruleList.forEach((rule, idx) => {
+        if (
+          indexesOfRulesThatMatter.indexOf(idx) > -1 &&
+          rule.rule !== identityRule
+        ) {
+          rule.matters = true;
+        }
+      });
+
+      if (reasons.length !== 1) {
+        // If identity wasn't the cause of the inequivalence, then one or more
+        // of the matters rules are the cause, so find out which one(s).
+        const _powerSet = powerSet(indexesOfRulesThatMatter);
```

Comment:
@mkalish and I discussed this power set thing a bit in the office after our meeting and a couple things came up that seemed like good reasons to drop it for now.

1. One of these rule combinations is essentially a new rule itself.
2. An equivalency with many rules could end up using a lot of CPU and causing performance problems on the client. e.g. 10 rules = 3,628,800 executions of the `for` on line 206 below.

This power set approach is a nice way to catch conflicting rules, however that can be handled through the standard API and the drawbacks seem to outweigh the benefit at the moment. It might be best to wait until we encounter a real-word problem where this would be needed before adding it in.

Reply:
So, it turns out that I was mistaken about not having a case where the power set is needed. It's actually the test case that is most relevant to our current work. The [cases where multiple rules are affecting the outcome](https://github.com/spanishdict/equivalency/blob/feature-reasons/index.test.js#L186-L206) won't be found unless the power set is checked, b/c it's only when the combination of rules that are affecting the outcome are changed from matters to doesn't matter that isEquivalent changes from false to true, allowing them to be discovered as the causes of the inequivalency.

The growth of the power set was something I was concerned about as well. A few notes about it:
- I believe the size of the power set is 2^length of array, so 10 rules would be 2^10=1024 sets to check
- it's opt-in -- reasons aren't calculated unless you ask for them, and for most of our equivalencies, such as correctness-checking, we won't ask for them. We'd only ask for them for incorrect answers.
- it only checks rules that matter, and usually there aren't many of those. For example, to check the things in our current epic, we only need three. I think before this, we've never used more than one of them in a particular equivalency instance.

Those assuaged my concerns -- I felt that the tradeoff of having this logic live in one place and being able to use the functionality of the library was worth it. If you're still concerned, one idea is to throw an error if reasons are asked for for an equivalency with more than, say, 16 matters rules. If we ever wanted to do that, we could then profile it to see how long it actually takes to execute -- might be acceptable, or if not, we could put effort into optimizing it at point.

Reply:
@sandinmyjoints I had missed the part that the power set only gets run through when determining the reasons for the inequality, which alleviates most of my concerns.  I do think putting a limit on the maximum rules that can be added would not hurt.

Reply:
Oh woops, you're right that the growth of the power set is 2^X, not X! ... That alone makes me feel better about keeping it. 

I was also confused about how the reasons code worked and I see now that it is necessary for the power set to exist in order to catch the multiple-rules case.

Thanks for breaking that down thoroughly -- this LGTM!

Reply:
OK, I added an option to explicitly opt-in to getting reasons for >16 matters rules, otherwise it will throw. I updated the docstring accordingly.

Reply:
On my machine, 16 rules takes ~75 ms, 22 rules takes ~8 seconds, and 23 rules throws an out of memory error.